kernel/pci/src/lib.rs

```diff
@@ -881,25 +905,45 @@ impl PciDevice {
     }
 
     /// Enables/Disables legacy (INTx) interrupts for this device
-    pub fn pci_enable_interrupts(&self, enable: bool) {
-        self.pci_set_interrupt_disable_bit(!enable);
+    pub fn pci_enable_intx(&self, enable: bool) {
+        self.pci_set_intx_disable_bit(!enable);
     }
 
-    /// Reads and returns this PCI device's interrupt status flag.
-    pub fn pci_get_interrupt_status(&self, check_enabled: bool) -> bool {
-        const PCI_STATUS_INT: u16 = 1 << 3;
+    /// Checks that legacy interrupts are enabled in the command register.
+    pub fn pci_intx_enabled(&self) -> bool {
+        (self.pci_read_16(PCI_COMMAND) & PCI_COMMAND_INT_DISABLED) == 0
+    }
 
-        let interrupt_enabled = || (self.pci_read_16(PCI_COMMAND) & PCI_COMMAND_INT_DISABLED) == 0;
-        let pending_interrupt = || (self.pci_read_16(PCI_STATUS)  & PCI_STATUS_INT          ) != 0;
+    /// Reads and returns this PCI device's legacy interrupt status flag.
+    pub fn pci_get_intx_status(&self, check_enabled: bool) -> bool {
+        const PCI_STATUS_INT: u16 = 1 << 3;
+        let pending_interrupt = || (self.pci_read_16(PCI_STATUS) & PCI_STATUS_INT) != 0;
 
-        ((!check_enabled) || interrupt_enabled()) && pending_interrupt()
+        ((!check_enabled) || self.pci_intx_enabled()) && pending_interrupt()
     }
 
-    /// Sets a task waker to be used when this device triggers an interrupt
+    /// Sets a task waker to be used when this device triggers a legacy interrupt
     ///
     /// Returns the previous interrupt waker for this device, if there was one.
-    pub fn set_interrupt_waker(&'static self, waker: Waker) -> Option<Waker> {
-        let mut handle = self.interrupt_waker.lock();
+    pub fn set_intx_waker(&'static self, waker: Waker) -> Option<Waker> {
+        #[cfg(target_arch = "x86_64")] {
+            // on x86, make sure that we register our handler for this IRQ
+            let int_num = match self.pci_get_intx_info() {
+                Ok((Some(irq), _pin)) => (irq + IRQ_BASE_OFFSET) as InterruptNumber,
+                _ => panic!("Cannot use INTx on device {:?}: No INTx info", self),
+            };
+
+            if let Err(existing_handler) = register_interrupt(int_num, pci_intx_handler) {
+                if existing_handler != (pci_intx_handler as usize) {
+                    panic!("Couldn't set PCI INTx handler for device: {:?}", self);
+                }
+            }
+
+            // on aarch64, this is done earlier (see the public init function)
```

Comment:
this is quite a confusing workflow that spans multiple functions. Can't we unify it to be the same across both x86 and aarch64?

Reply:
what's "hacky" about the x86 solution? You mean the way we "discover" (or assume) the IRQ numbers on x86 is hacky?

My comment here was just about the way that x86 and aarch64 initialization is split across multiple functions in a way that makes it really hard to follow, which means it is hard to maintain. From a design perspective, it'd be nice if a single function or a set of functions handled both aarch64 and x86 initialization using the same pattern/structure of code.